substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/isolated/IsolatedSpeculationLog.java

```diff
@@ -48,19 +58,115 @@ public void collectFailedSpeculations() {
         collectFailedSpeculations0(IsolatedCompileContext.get().getClient(), handle);
     }
 
+    static class SpeculationReasonInvocationHandler implements InvocationHandler {
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            return IsolatedSpeculationReasonEncoding.class.getDeclaredMethod(method.getName(), method.getParameterTypes()).invoke(args);
+        }
+    }
+
+    static class SpeculationReasonEncodingSupplier implements Supplier<Object> {
+        private final Object speculationReasonEncodingProxy;
+
+        SpeculationReasonEncodingSupplier(Object speculationReasonEncodingProxy) {
+            this.speculationReasonEncodingProxy = speculationReasonEncodingProxy;
+        }
+
+        @Override
+        public Object get() {
+            try {
+                return speculationReasonEncodingProxy.getClass().getDeclaredConstructor().newInstance();
```

Comment:
Probably best to just do `PROXY_CLASS.getDeclaredConstructor(...).newInstance()` here so the native image build can fold the reflective instantiation.

Reply:
I cached the constructor as well.